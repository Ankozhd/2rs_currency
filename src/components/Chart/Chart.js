import React, { useEffect, useState } from 'react';
import { currencyService } from '../../services/currency.service';
// Dropdown
import ReactDropdown from 'react-dropdown';
import 'react-dropdown/style.css';

// Date Picker
import DatePicker from 'react-datepicker';

// Workaround for JavaScript heap out of memory error with plotly
// https://github.com/plotly/react-plotly.js/issues/135

import Plotly from 'plotly.js-basic-dist';

import createPlotlyComponent from 'react-plotly.js/factory';
// Components
import Utils from '../../utils/Utils';
import Stats from '../Stats/Stats';

const Plot = createPlotlyComponent(Plotly);

function Chart({
  base, symbol, symbols, setCompareSymbol
}) {
  const [state, setState] = useState({
    xAxis: [],
    yAxis: []
  });

  const [startDate, setStartDate] = useState(new Date().setDate(new Date().getDate() - 7));
  const [endDate, setEndDate] = useState(new Date());

  useEffect(() => {
    updateChart(base, symbol, Utils.dateToEuropean(startDate), Utils.dateToEuropean(endDate));
  }, [base, symbol, startDate, endDate]);

  function updateChart(base, symbol, start, end) {
    currencyService.getHistoricalRatesForPeriod(base, symbol, start, end).then(
      (data) => {
        // In ideal world all these mutations have to be done on server side
        // Front should just receive xAxis and yAxis arrays

        // Get property names as x axis values
        const xAxis = Object.keys(data.rates);
        const yAxis = [];

        // Object keys do not guarantee correct order
        // Sorting manually
        xAxis.sort();

        // Find corresponding value for each x axis
        xAxis.forEach((date) => {
          yAxis.push(data.rates[date][symbol]);
        });

        setState({ xAxis, yAxis });
      }
    );
  }

  return (
    <div>
      <div>Compare to: {symbol}</div>
      <div>
        <ReactDropdown
          options={symbols.filter((smb) => smb !== base)}
          onChange={(e) => setCompareSymbol(e.value)}
          value={symbol}
          placeholder="Select symbol to compare against"
        />
      </div>
      <div>Time Span: {Utils.dateToEuropean(startDate)} -> {Utils.dateToEuropean(endDate)}</div>
      <div className="d-flex">
        <div className="d-flex flex-column justify-content-between mr-3">
          <div className="d-flex justify-content-between">
            <DatePicker
              dateFormat="yyyy/MM/dd"
              selected={startDate}
              onChange={(date) => setStartDate(date)}
              selectsStart
              startDate={startDate}
              endDate={endDate}
              maxDate={new Date()}
            />
            <DatePicker
              dateFormat="yyyy/MM/dd"
              selected={endDate}
              onChange={(date) => setEndDate(date)}
              selectsEnd
              startDate={startDate}
              endDate={endDate}
                            // minDate={startDate}
              maxDate={new Date()}
            />
          </div>
          <Plot
            data={[
              {
                x: state.xAxis,
                y: state.yAxis,
                type: 'scatter',
                mode: 'lines+markers',
                marker: { color: 'blue' }
              }
            ]}
            layout={{ width: 640, height: 480, title: `${base} to ${symbol}` }}
          />
        </div>

        <Stats prices={state.yAxis} />
      </div>
    </div>
  );
}

export default Chart;
