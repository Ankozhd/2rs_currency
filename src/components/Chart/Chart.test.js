import React from 'react';
import { render } from '@testing-library/react';
import Chart from './Chart';

describe('<Chart />', () => {
  const { container, getByText } = render(<Chart base="USD" symbol="EUR" symbols={['USD', 'EUR', 'CAD']} setCompareSymbol={() => {}} />);
  it('should render symbol correctly', () => {
    expect(getByText('Compare to: EUR')).toBeInTheDocument();
  });
});
