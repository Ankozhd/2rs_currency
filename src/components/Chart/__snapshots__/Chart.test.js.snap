// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> should render and match the snapshot 1`] = `
<div>
  <div>
    Compare to: 
    EUR
  </div>
  <div>
    <Dropdown
      baseClassName="Dropdown"
      onChange={[Function]}
      options={
        Array [
          "EUR",
          "CAD",
        ]
      }
      placeholder="Select symbol to compare against"
      value="EUR"
    />
  </div>
  <div>
    Time Span: 
    2020-05-29
     -&gt; 
    2020-06-05
  </div>
  <div>
    <a
      allowSameDay={false}
      customTimeInput={null}
      dateFormat="yyyy/MM/dd"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      enableTabLoop={true}
      endDate={2020-06-05T05:43:08.813Z}
      excludeScrollbar={true}
      focusSelectedMonth={false}
      maxDate={2020-06-05T05:43:08.814Z}
      monthsShown={1}
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onCalendarClose={[Function]}
      onCalendarOpen={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1590730988813}
      selectsStart={true}
      shouldCloseOnSelect={true}
      showFullMonthYearPicker={false}
      showMonthYearPicker={false}
      showPopperArrow={true}
      showPreviousMonths={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      showTwoColumnMonthYearPicker={false}
      showYearPicker={false}
      startDate={1590730988813}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <a
      allowSameDay={false}
      customTimeInput={null}
      dateFormat="yyyy/MM/dd"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      enableTabLoop={true}
      endDate={2020-06-05T05:43:08.813Z}
      excludeScrollbar={true}
      focusSelectedMonth={false}
      maxDate={2020-06-05T05:43:08.814Z}
      monthsShown={1}
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onCalendarClose={[Function]}
      onCalendarOpen={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected={2020-06-05T05:43:08.813Z}
      selectsEnd={true}
      shouldCloseOnSelect={true}
      showFullMonthYearPicker={false}
      showMonthYearPicker={false}
      showPopperArrow={true}
      showPreviousMonths={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      showTwoColumnMonthYearPicker={false}
      showYearPicker={false}
      startDate={1590730988813}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
  </div>
  <PlotlyComponent
    data={
      Array [
        Object {
          "marker": Object {
            "color": "blue",
          },
          "mode": "lines+markers",
          "type": "scatter",
          "x": Array [],
          "y": Array [],
        },
      ]
    }
    debug={false}
    layout={
      Object {
        "height": 480,
        "title": "USD to EUR",
        "width": 640,
      }
    }
    style={
      Object {
        "display": "inline-block",
        "position": "relative",
      }
    }
    useResizeHandler={false}
  />
  <Stats
    prices={Array []}
  />
</div>
`;
